### PR TITLE
13 variable sysmods

### PR DESCRIPTION
**Pull request by CompBioDan**
# Requester section

*To be completed by the pull-requester (assignee)*

## Description
<details><summary>Please include a summary of the changes</summary>
 
  - What is this change?
  - What does it fix?
  - Is this a bug fix or a feature and does it break any existing functionality?
  - How has it been tested?
</details>
 
This pull request adds the functionality to include/exclude specific individual system models from the fitting and estimation. It achieves this by modifying the objective function .cpp file (src/dpmaccount.cpp) to accept an additional include_sysmods vector of flags (0/1) with the expected format of [births, deaths, ins, outs] (e.g. [1, 1, 0, 0] to include births/deaths and exclude ins/outs) which determines whether each system model contributes to the estimation of the negative log posterior. The upstream functions (R/comod-methods.R (**fit.dpmaccount()**) and  R/estimate_account.R **(estimate_account()**) have been modified to accept/pass through this user-defined flag vector from estimate_account(), through the fit.dpmaccount() and into the modified .cpp objective function. 
-  This implementation therefore still requires the system models to be specified/provided, to limit the changes to the system model specific checks elsewhere in the package. It allows us to easily include/exclude the system models without requiring additional potentially breaking changes. 
- This makes the additional functionality non-breaking, as the default including all system models remains the same as the previous version 0.3.1 and all results when including system models in 3.1 and this 3.2 should be the same/reproducible. 
- Unit tests have been added to check both the fit and the estimate_account function (that fit produces non-NA estimates, and that estimate_account works with the new specification). 

 
## Type of change
- [ ] Bug fix - *non-breaking change*
- [X] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...
 
## Requestor checklist:

*If applicable, if not indicate why in requester comments section*

- [X] I have performed a self-review of my own code
- [X] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how)
- [X] I have made corresponding changes to the documentation (comments, docstring, etc.) using roxygen (devtools check()/document() etc.)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have updated the change log/version number [CHANGELOG.md](https://github.com/ONSdigital/dpmaccount/blob/main/CHANGELOG.md)
- [X] I have checked the pipeline runs with test data

### Requester comments
 
*Any additional requester comments/notes/requirements*

<br>
 
# Reviewer section

*To be completed by the pull-request reviewer*

##  Peer review checklist:
Any new code (where appropriate) includes all of the following:

- [ ] **Commit messages**:
  - [ ] The individual commit messages are sufficiently descriptive, when combined with the **pull request description** above, forming a coherent audit trail. If not, add suggested improvements/changes to the reviewer notes below
  - [ ] The commit messages are in-line with the [contributing guidance](https://github.com/ONSdigital/dpmaccount/blob/main/CONTRIBUTING.md)
- [ ] **Data**: All [personally identifiable data](https://ico.org.uk/for-organisations/uk-gdpr-guidance-and-resources/personal-information-what-is-it/what-is-personal-information-a-guide/) has been removed/anonymised
- [ ] **Documentation**: Docstrings, comments have been added/updated
- [ ] **[Style guidelines](https://confluence.ons.gov.uk/display/PP/Style+guidelines)**: New code conforms to the project's contribution guidelines
- [ ] **[Functionality](https://confluence.ons.gov.uk/display/PP/Functionality)**: The code works fully implements the requirements and works as expected, handles expected edge cases, exceptions are handled appropriately
- [ ] **[Complexity](https://confluence.ons.gov.uk/display/PP/Complexity)**: The code is not overly complex, logic has been split into appropriately sized functions, etc.
- [ ] **[Test coverage](https://confluence.ons.gov.uk/display/PP/Test+coverage)**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.
- [ ] **Vignettes**: Any relevant vignettes have been modified/created.
- [ ] **Version numbering**: The versioning number has been incremented correctly and follows mandated semantic versioning protocol in [contributing guidance](https://github.com/ONSdigital/dpmaccount/blob/main/CONTRIBUTING.md).

*If any of the above are not applicable, include an explanation in the review comments below*

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>
 
- bugs that need fixing (does it work as expected? and does it work with other code that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
- Do the tests effectively assure that it works correctly? Are there additional edge cases/negative tests to be considered?
- Code style improvements (could the code be written more clearly?)
</details>
<br>
 
*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: data-raw/sysmod.R
Comment (lmorris2) [inline data-raw/sysmod.R:1]: approve - white-space removal only

File: R/check-functions.R
Comment (lmorris2) [inline R/check-functions.R:1]: approve - white-space removal only

File: R/codatamod-methods.R
Comment (lmorris2) [inline R/codatamod-methods.R:1]: approve - white-space removal only

File: R/results-methods.R
Comment (lmorris2) [inline R/results-methods.R:1]: approve - white-space removal only

File: R/util.R
Comment (lmorris2) [inline R/util.R:1]: approve - white-space removal only

File: tests/testthat/test-check-functions.R
Comment (lmorris2) [inline tests/testthat/test-check-functions.R:1]: approve - white-space removal only

File: tests/testthat/test-codatamod-constructors.R
Comment (lmorris2) [inline tests/testthat/test-codatamod-constructors.R:1]: approve - white-space removal only

File: tests/testthat/test-codatamod-methods.R
Comment (lmorris2) [inline tests/testthat/test-codatamod-methods.R:1]: approve - white-space removal only

File: tests/testthat/test-components-helpers.R
Comment (lmorris2) [inline tests/testthat/test-components-helpers.R:1]: approve - white-space removal only

File: tests/testthat/test-results-methods.R
Comment (lmorris2) [inline tests/testthat/test-results-methods.R:1]: approve - white-space removal only

File: DESCRIPTION
Comment (lmorris2) [inline DESCRIPTION:1]: approve

File: man/estimate_account.Rd
Comment (lmorris2) [inline man/estimate_account.Rd:1]: approve

File: man/fit.dpmaccount_comod.Rd
Comment (lmorris2) [inline man/fit.dpmaccount_comod.Rd:1]: approve

File: R/comod-methods.R [hunk 1]
```diff
@@ -80,7 +84,8 @@ fit.dpmaccount_comod <- function(object, keep_adfun = FALSE, ...) {
     i_mod_all_outs = get_i_mod_all(datamods_outs),
     has_par_all_stk = get_has_par_all(datamods_stk),
     has_par_all_ins = get_has_par_all(datamods_ins),
-    has_par_all_outs = get_has_par_all(datamods_outs)
+    has_par_all_outs = get_has_par_all(datamods_outs),
```
Comment (lmorris2) [inline R/comod-methods.R:87]: are these unit tests? do you want one on include_sysmods to check that vals are 0 or 1? what happens if it fails? do you get an error message? I did a quick scan of the pull req and didn't see one written in

File: R/estimate_account.R
Comment (lmorris2) [inline R/estimate_account.R:1]: approve

File: src/dpmaccount.cpp
Comment (lmorris2) [inline src/dpmaccount.cpp:1]: looks like the previous system models packaged up into code which switches them on or off so looks ok to me

File: tests/testthat/test-comod-methods.R
Comment (lmorris2) [inline tests/testthat/test-comod-methods.R:1]: not sure I fully understand but looks like a test to see if excluding all sysmods still generates the expected model object and produces results? if so all good I think?

File: tests/testthat/test-estimate_account.R
Comment (lmorris2) [inline tests/testthat/test-estimate_account.R:1]: Is there a reason to have two tests here but not in the comod script? Also why are the tests here slightly different?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,15 @@
 # Pre-release changes/updates
 
+# dpmaccount 0.3.2
+
+## Implemented variable system model inclusion
+
+- Modified the objective function (dpmaccount.cpp) to allow variable system model inclusion based on a new 'include_sysmods' flag vector. This flag vector indicates whether to include (1) or exclude (0) each of the system models (c(births, deaths, ins, outs), e.g. c(1, 1, 0, 0) includes births and deaths system models, excludes ins and out system models).
```
Comment (lmorris2) [inline CHANGELOG.md]:
```suggestion
- Modified the objective function (dpmaccount.cpp) to allow variable system model inclusion based on a new 'include_sysmods' flag vector. This flag vector indicates whether to include (1) or exclude (0) each of the system models (c(births, deaths, ins, outs), e.g. c(1, 1, 0, 0) includes births and deaths system models, excludes ins and outs system models). 
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
 
+# dpmaccount 0.3.2
+
+## Implemented variable system model inclusion
+
+- Modified the objective function (dpmaccount.cpp) to allow variable system model inclusion based on a new 'include_sysmods' flag vector. This flag vector indicates whether to include (1) or exclude (0) each of the system models (c(births, deaths, ins, outs), e.g. c(1, 1, 0, 0) includes births and deaths system models, excludes ins and out system models). 
+
+- Modified `fit.dpmaccount_comod` method (comod-methods.R) to accept the 'include_sysmods' flag vector and pass it to the objective function through the MakeADFun() method.
+
+- Modified `estimate_account()` (estimate_account.R) to accept the new 'include_sysmods' flag vector input, in the format c(births, deaths, ins, outs), e.g. c(1, 1, 0, 0) includes births and deaths system models, excludes ins and out system models.
```
Comment (lmorris2) [inline CHANGELOG.md]:
```suggestion
- Modified `estimate_account()` (estimate_account.R) to accept the new 'include_sysmods' flag vector input, in the format c(births, deaths, ins, outs), e.g. c(1, 1, 0, 0) includes births and deaths system models, excludes ins and outs system models. 
```

File: R/comod-methods.R (hunk 1 above)
Comment (CompBioDan) [reply R/comod-methods.R:87]: These specific functions are used to switch between the parameterised TMB data models (e.g. to use _scale_ratio_ or not) but you make a very valid point we definitely want to have checks on _include_sysmods_ before it gets passed through _estimate_account_ to the _fit_ method. This wasn't previously included, but I have added one now (added _check_include_sysmods()_ to _estimate_account.R_ to sit alongside the other input checks occurring right at the start of the process to prevent unnecessary running of downstream code doomed to fail later on.

File: src/dpmaccount.cpp
Comment (CompBioDan) [reply src/dpmaccount.cpp:1]: That's exactly the extent of the changes to the objective function code, to accept a new flag vector and to use its contents to switch on/off the inclusion of each system model in the estimation of the posterior.

File: tests/testthat/test-comod-methods.R
Comment (CompBioDan) [reply tests/testthat/test-comod-methods.R:1]: Yes so the test is specifically designed to just make sure that the _fit_ method still works when we exclude system models, in that it produces the correct object (a _dpmaccount_comod_ object) and that the mean isn't NA. I have now added a second check for the _fit_ method for logical inclusion flags (c(TRUE, TRUE, FALSE, FALSE)) in response to your valid point about checking estimate_account for both but not fit.

File: tests/testthat/test-estimate_account.R
Comment (CompBioDan) [reply tests/testthat/test-estimate_account.R:1]: This is a good point, as a more downstream function it would make more sense to check _fit_ for both integer and logical flags over checking _estimate_account_, so I've added the second test to the _fit_ unit tests now. It won't hurt to keep this unit test for estimate_account incase future changes somehow break this function without breaking _fit_.

**Comment by CompBioDan** [comment]
Changes in commit https://github.com/ONSdigital/dpmaccount/pull/14/changes/9b79f05bb3faf4f421bd51f547cf0f19a9dc8543 introduce a new data validation function, check_include_sysmods, along with a testthat suite, to make sure that we are correctly handling the system model inclusion flags before we use them in the account estimation. Change in https://github.com/ONSdigital/dpmaccount/pull/14/changes/2e26802b4fdaa26cf73293695ded78b0f42bd1b3 checks for valid handling of logical flags in include_sysmods provided to _fit_ method.
Key Changes:
•	Check_include_sysmods(): Added a function that strictly enforces that the include_sysmods vector contains exactly 4 values restricted to 0, 1, TRUE, or FALSE
•	Compound Error Reporting: Implemented user-friendly error handling that evaluates both length and content together. If an input fails, it collects all issues and returns a single, formatted bulleted list (e.g., printing the exact length received and extracting the specific invalid values). His prevents the user from hitting a length error, fixing it, and immediately hitting a value error. 
•	Pipeline Compatibility: The function returns the input indivisibly upon success, making it safe to use within %>% or |> pipelines.
•	Unit Tests: Added a full suite of tests in tests/testthat/ covering valid passes, invalid lengths, invalid data typers/values (including NA), and compound errors.

File: R/check-functions.R
Comment (lmorris2) [inline R/check-functions.R:1]: looks good as far as I can tell a detailed unit test for the include_sysmods function as I previously suggested

File: R/comod-methods.R (hunk 1 above)
Comment (lmorris2) [reply R/comod-methods.R:87]: can see check-functions.R contains the unit test I suggested

File: R/estimate_account.R
Comment (lmorris2) [inline R/estimate_account.R:1]: unit test on include_sysmods successfully added

File: src/dpmaccount.cpp
Comment (lmorris2) [reply src/dpmaccount.cpp:1]: great

File: tests/testthat/test-check-functions.R [hunk 4]
```diff
@@ -1734,3 +1733,71 @@ test_that("'check_time_complete' returns correct error with invalid inputs", {
...
+
+  # Missing values (NA)
+  expect_error(
+    check_include_sysmods(c(1, 0, NA, 1)),
+    "Found invalid values: NA"
+  )
+
+  # Characters/Strings
+  expect_error(
+    check_include_sysmods(c("TRUE", "FALSE", "TRUE", "FALSE")),
```
Comment (lmorris2) [inline tests/testthat/test-check-functions.R:1787]: Is there a discrepancy here because https://github.com/ONSdigital/dpmaccount/pull/14/changes#diff-e8d75f8eb95719b08d3063d7099008e3a885c1b1311e1f7dfd3a6be1681a46a6R1465 suggests that TRUE and FALSE are acceptable values.

File: tests/testthat/test-check-functions.R
Comment (lmorris2) [inline tests/testthat/test-check-functions.R:1]: other than the TRUE/FALSE thing looks acceptable

File: tests/testthat/test-comod-methods.R
Comment (lmorris2) [reply tests/testthat/test-comod-methods.R:1]: see this comment https://github.com/ONSdigital/dpmaccount/pull/14/changes#r3767767999 I think there is a discrepency with that unit test there which suggests TRUE/FALSE values are a fail

Comment (lmorris2) [reply tests/testthat/test-comod-methods.R:1]: otherwise the second test here seems to be fine

File: tests/testthat/test-estimate_account.R
Comment (lmorris2) [reply tests/testthat/test-estimate_account.R:1]: sounds fine

**Review by lmorris2 — COMMENTED** [review]
I think there is a discrepancy where one of the unit tests suggests that TRUE/FALSE values in the include_sysmod function are unacceptable whereas other scripts suggest they are accetpable

File: tests/testthat/test-check-functions.R (hunk 4 above)
Comment (CompBioDan) [reply tests/testthat/test-check-functions.R:1787]: Ah so I've caused some confusion with my choice of character strings, the test should fail when you provide any strings (including "TRUE"/"FALSE" but it'll pass with the logical factors (TRUE/FALSE). It was the first example I could think of but I realise now how misleading it could look.

Comment (lmorris2) [reply tests/testthat/test-check-functions.R:1787]: Makes sense! Can you change it to something else? Then happy to pass to Andrew or Aidan etc for double check.
